be/src/exec/vectorized/chunks_sorter_full_sort.cpp

```diff
@@ -2,159 +2,114 @@
 
 #include "chunks_sorter_full_sort.h"
 
-#include "column/type_traits.h"
+#include "exec/vectorized/sorting/merge.h"
 #include "exec/vectorized/sorting/sort_permute.h"
 #include "exec/vectorized/sorting/sorting.h"
 #include "exprs/expr.h"
-#include "gutil/casts.h"
-#include "runtime/primitive_type_infra.h"
 #include "runtime/runtime_state.h"
-#include "util/orlp/pdqsort.h"
 #include "util/stopwatch.hpp"
 
 namespace starrocks::vectorized {
 
 ChunksSorterFullSort::ChunksSorterFullSort(RuntimeState* state, const std::vector<ExprContext*>* sort_exprs,
                                            const std::vector<bool>* is_asc, const std::vector<bool>* is_null_first,
                                            const std::string& sort_keys)
-        : ChunksSorter(state, sort_exprs, is_asc, is_null_first, sort_keys, false) {
-    _selective_values.resize(_state->chunk_size());
-}
+        : ChunksSorter(state, sort_exprs, is_asc, is_null_first, sort_keys, false) {}
 
 ChunksSorterFullSort::~ChunksSorterFullSort() = default;
 
 Status ChunksSorterFullSort::update(RuntimeState* state, const ChunkPtr& chunk) {
-    if (UNLIKELY(_big_chunk == nullptr)) {
-        _big_chunk = chunk->clone_empty();
-    }
+    _merge_unsorted(state, chunk);
+    _partial_sort(state, false);
 
-    size_t target_rows = _big_chunk->num_rows() + chunk->num_rows();
-    if (target_rows > Column::MAX_CAPACITY_LIMIT) {
-        LOG(WARNING) << "Full sort rows exceed limit " << target_rows;
-        return Status::InternalError(fmt::format("Full sort rows exceed limit: {}", target_rows));
-    }
+    return Status::OK();
+}
 
-    _big_chunk->append(*chunk);
+// Accumulate unsorted input chunks into a larger chunk
+Status ChunksSorterFullSort::_merge_unsorted(RuntimeState* state, const ChunkPtr& chunk) {
+    SCOPED_TIMER(_build_timer);
 
-    if (_big_chunk->reach_capacity_limit()) {
-        LOG(WARNING) << "Full sort encounter big chunk overflow issue";
-        return Status::InternalError(fmt::format("Full sort encounter big chunk overflow issue"));
+    if (_unsorted_chunk == nullptr) {
+        // TODO: optimize the copy
+        _unsorted_chunk.reset(chunk->clone_unique().release());
+    } else {
+        _unsorted_chunk->append(*chunk);
     }
 
-    DCHECK(!_big_chunk->has_const_column());
     return Status::OK();
 }
 
-Status ChunksSorterFullSort::done(RuntimeState* state) {
-    if (_big_chunk != nullptr && _big_chunk->num_rows() > 0) {
-        RETURN_IF_ERROR(_sort_chunks(state));
+// Sort the large chunk
+Status ChunksSorterFullSort::_partial_sort(RuntimeState* state, bool done) {
+    if (!_unsorted_chunk) {
+        return Status::OK();
+    }
+    bool reach_limit = _unsorted_chunk->num_rows() >= kMaxBufferedChunkSize || _unsorted_chunk->reach_capacity_limit();
+    if (done || reach_limit) {
+        SCOPED_TIMER(_sort_timer);
+
+        // Check column overflow problem
+        RETURN_IF_ERROR(_unsorted_chunk->upgrade_if_overflow());
```

Comment:
if we call `upgrade_if_overflow` method, do we need to call `downgrade` method tool？

Reply:
As I know, the usage of `downgrade` is a big chunk of `LargeBinaryColumn` is split into small chunks again. But the full sort would not do this, its behavior is just:
1. Merge input chunks into a big chunk, check overflow before do partial sorting
2. Do merge sort, but not merge all chunks into a bigger chunk, keep them as `SortedRuns` instead.

Reply:
After discussion, the upgrade is removed, replace it with a capacity check, throw an exception if exceed limit.